Integrate with the Bolt registry to check if pubkeys are registered

thedevbirb opened the pull request:
Closes #332.

Introduces new checks for both the commitment signer key and the validator public keys which should correspond to an active operator and validator respectively in the `BoltManager` contract.

The check is performed at startup, when creating the `SidecarDriver::from_components`. This is a good spot because both signers (ECDSA, BLS) are abstracted via traits so the check can be safely done in a single pace.

 

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Introduced `public_key` method in the `SignerECDSA` trait and its implementations, allowing retrieval of the public key as an address.
	- Added `BoltManager` for enhanced key verification in the `SidecarDriver`.
	- Launched the `onchain_registry` module, expanding on-chain management capabilities.
	- Implemented various utility methods in `BoltManager` for address retrieval and operator/validator verification.

- **Bug Fixes**
	- Improved readability of conditional statements and return formats in several methods.

- **Documentation**
	- Enhanced comments for clarity in the `make_path` and `test_keystore_signer` functions.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

thedevbirb committed:
feat(sidecar): port BoltManager wrapper struct for validation purposes

thedevbirb committed:
feat(sidecar): add `isOperator` function to BoltManager wrapper

thedevbirb committed:
feat(sidecar): public_keys method on SignerECDSA trait

thedevbirb committed:
feat(sidecar): more utils to BoltManager + BoltValidators

thedevbirb committed:
feat(sidecar): verify operator and validator keys on startup

thedevbirb committed:
chore(sidecar): fmt

merklefruit commented on bolt-sidecar/src/onchain_registry.rs:
```diff
@@ -0,0 +1,163 @@
+pub struct BoltManager(BoltManagerContractInstance<Http<Client>, RootProvider<Http<Client>>>);
+
+impl BoltManager {
+    /// Returns the address of the canonical BoltManager contract for a given chain, if present
+    pub fn address(chain: Chain) -> Option<Address> {
+        match chain {
+            Chain::Holesky => Some(
+                Address::from_str("0x440202829b493F9FF43E730EB5e8379EEa3678CF")
+                    .expect("valid address"),
+            ),
+            _ => None,
+        }
+    }
```
why did you choose to place this here instead of in config::Chain where we already have all the chain specific stuff? To me it seems like a more appropriate place, wdyt?

mempirate replied:
Agreed

thedevbirb replied:
Ok, I don't have a strong preference I first imagined this as an association Manager -> Chain rather than Chain -> Manager but the latter works well too

mempirate commented on bolt-sidecar/src/onchain_registry.rs:
Wdyt about putting this in its own `chain_io` module with a file per contract? I.e. manager & validators

mempirate commented on bolt-sidecar/src/driver.rs:
```diff
@@ -161,6 +161,16 @@ impl<C: StateFetcher, ECDSA: SignerECDSA> SidecarDriver<C, ECDSA> {
+            bolt_manager.verify_operator(commitment_signer.public_key()).await?;
+            bolt_manager
+                .verify_validator_pubkeys(&Vec::from_iter(constraint_signer.available_pubkeys()))
+                .await?;
```
I think here we should also verify that the link between the validators and the operator is correct, i.e. the provided operator should be active for those validators

thedevbirb replied:
Right, great catch thank you

mempirate commented on bolt-sidecar/src/onchain_registry.rs:
```diff
@@ -0,0 +1,163 @@
+    }
+}
+
+sol! {
```
In the future we should probably do something to keep these interfaces up to date, @coderabbitai please draft an issue

mempirate replied:
Beast

mempirate commented:
@coderabbitai review

thedevbirb committed:
refactor(sidecar): chain_io module

thedevbirb committed:
fix(sidecar): ensure match between commitment signer public key and validators' authorized operator

thedevbirb committed:
chore(ci/linkinspector): add ignore pattern for etherscan links

Etherscan links usually does some checks like user agent for dos
protection, to avoid that we should call the api with a proper API key

thedevbirb committed:
fix(sidecar): add more descriptive error when querying contracts

ValidatorDoesNotExist now returns also the public key hash

thedevbirb committed:
chore(sidecar/chain_io): better error message, drop isOperator check since getProposerStatuses does that already

thedevbirb committed:
fix(sidecar): use compressed hash bytes20 instead of bytes32

thedevbirb committed:
test(sidecar): bolt manager wrapper contract

merklefruit commented on bolt-sidecar/src/chain_io/manager.rs:
```diff
@@ -0,0 +1,169 @@
+
+    use super::BoltManager;
+
+    #[tokio::test]
```
Can we add #[ignore] here? Otherwise CI won't pass

merklefruit commented on bolt-sidecar/src/chain_io/utils.rs:
Very nice comments + tests, great job

merklefruit reviewed (approved):
Great!

thedevbirb committed:
chore(sidecar): update links

mempirate commented on bolt-sidecar/src/config/chain.rs:
```diff
@@ -102,6 +104,17 @@ impl Chain {
+    pub fn manager_address(&self) -> Option<Address> {
+        match self {
+            Chain::Holesky => Some(
+                Address::from_str("0x440202829b493F9FF43E730EB5e8379EEa3678CF")
```
Can we put this in a constant / static?

mempirate reviewed (changes requested):
Last nit, great work

thedevbirb committed:
chore(sidecar): const manager address
